### PR TITLE
[Feature] Disabling custom fields

### PR DESCRIPTION
**Pull request by Civolilah**
Here's a screenshot of the implemented solution to disable input for custom fields for users who don't have access to them:

Screenshot 2022-12-28 at 00 48 00

### CODE REVIEW

**Comment by turbo124** [comment]
Please use the following translation:

upgrade_to_paid_plan

**Comment by Civolilah** [comment]
> 

@turbo124 The keyword has been changed. Here is the screeshot of how it looks now:

Screenshot 2022-12-28 at 00 48 00

**Comment by turbo124** [comment]
@Civolilah @beganovich LGTM

File: src/components/CustomFieldsPlanAlert.tsx [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+  const [t] = useTranslation();
+
+  const user = useCurrentUser();
+
+  const showAlert = !proPlan() && !enterprisePlan() && isHosted();
+
+  return (
+    <>
+      {showAlert && (
+        <div className={`${props.className}`}>
```
Comment (beganovich) [inline src/components/CustomFieldsPlanAlert.tsx:31]: No need to encapsulate the variable here.

```ts
<div className={props.className}>
```

should work just fine.

File: src/pages/payments/edit/PaymentFields.tsx [hunk 2]
```diff
@@ -9,61 +9,86 @@
...
-      </Element>
+        <Element
+          leftSide={
+            <div className="inline-flex items-center space-x-2">
+              <span>{t('note')}</span>
+              <span className="text-red-600">*</span>
+            </div>
+          }
+        >
+          Custom fields apply to all Payments, they are not specific to this
```
Comment (beganovich) [inline src/pages/payments/edit/PaymentFields.tsx:76]: Can we translate this string?

File: src/pages/products/edit/ProductFields.tsx [hunk 3]
```diff
@@ -9,63 +9,89 @@
...
+
   const dispatch = useDispatch();
 
-  const [t] = useTranslation();
+  const company = useInjectCompanyChanges();
+
   const handleCustomFieldChange = useHandleCustomFieldChange();
 
-  const onSave = (event: React.FormEvent<HTMLFormElement>) => {
+  const disabledCustomFields = !proPlan() && !enterprisePlan() && isHosted();
```
Comment (beganovich) [inline src/pages/products/edit/ProductFields.tsx:35]: Looks like we are copying the logic for disabling custom fields each time.
Can we move this into a hook, so we have logic centralized?

File: src/pages/products/edit/ProductFields.tsx [hunk 4]
```diff
@@ -9,63 +9,89 @@
...
-      </Element>
+        <Element
+          leftSide={
+            <div className="inline-flex items-center space-x-2">
+              <span>{t('note')}</span>
+              <span className="text-red-600">*</span>
+            </div>
+          }
+        >
+          Custom fields apply to all products, they are not specific to this
```
Comment (beganovich) [inline src/pages/products/edit/ProductFields.tsx:79]: Can we translate this string?

File: src/pages/invoices/common/components/InvoiceFooter.tsx [hunk 5]
```diff
@@ -54,6 +57,8 @@ export function InvoiceFooter(props: Props) {
   const handleCustomSurchargeFieldChange =
     useHandleCustomSurchargeFieldChange();
 
+  const disabledCustomFields = !proPlan() && !enterprisePlan() && isHosted();
```
Comment (beganovich) [inline src/pages/invoices/common/components/InvoiceFooter.tsx:60]: Move into hook.

File: src/pages/settings/custom-fields/components/Field.tsx [hunk 6]
```diff
@@ -35,6 +38,9 @@ export function Field(props: Props) {
   const [initialValue, setInitialValue] = useState('');
   const [dropdownInitialValue, setDropdownInitialValue] = useState('');
 
+  const disabledInputCustomFields =
+    !proPlan() && !enterprisePlan() && isHosted();
+
```
Comment (beganovich) [inline src/pages/settings/custom-fields/components/Field.tsx:43]: Move into hook.

File: src/pages/settings/custom-fields/invoices/Invoices.tsx [hunk 7]
```diff
@@ -18,11 +22,15 @@ import { Field } from '../components';
 
 export function Invoices() {
   const [t] = useTranslation();
+
+  const disabledCustomFields = !proPlan() && !enterprisePlan() && isHosted();
+
```
Comment (beganovich) [inline src/pages/settings/custom-fields/invoices/Invoices.tsx:27]: Move into hook.

File: src/pages/settings/custom-fields/invoices/Invoices.tsx [hunk 8]
```diff
@@ -18,11 +22,15 @@ import { Field } from '../components';
...
   const pages = [
     { name: t('settings'), href: '/settings' },
     { name: t('custom_fields'), href: '/settings/custom_fields' },
     { name: t('invoices'), href: '/settings/custom_fields/invoices' },
   ];
+
   useEffect(() => {
     document.title = `${import.meta.env.VITE_APP_TITLE}: ${t('custom_fields')}`;
   });
```
Comment (beganovich) [inline src/pages/settings/custom-fields/invoices/Invoices.tsx:36]: Can we refactor this to use `useTitle` hook?

File: src/pages/settings/custom-fields/payments/Payments.tsx [hunk 9]
```diff
@@ -16,12 +17,15 @@ import { Field } from '../components';
...
+
   const pages = [
     { name: t('settings'), href: '/settings' },
     { name: t('custom_fields'), href: '/settings/custom_fields' },
     { name: t('payments'), href: '/settings/custom_fields/payments' },
   ];
+
   useEffect(() => {
     document.title = `${import.meta.env.VITE_APP_TITLE}: ${t('custom_fields')}`;
   });
```
Comment (beganovich) [inline src/pages/settings/custom-fields/payments/Payments.tsx:31]: Can we refactor this to use `useTitle` hook?

File: src/pages/settings/custom-fields/products/Products.tsx [hunk 10]
```diff
@@ -16,12 +17,15 @@ import { Field } from '../components';
...
+
   const pages = [
     { name: t('settings'), href: '/settings' },
     { name: t('custom_fields'), href: '/settings/custom_fields' },
     { name: t('products'), href: '/settings/custom_fields/products' },
   ];
+
   useEffect(() => {
     document.title = `${import.meta.env.VITE_APP_TITLE}: ${t('custom_fields')}`;
   });
```
Comment (beganovich) [inline src/pages/settings/custom-fields/products/Products.tsx:31]: Can we refactor this to use `useTitle` hook?

File: src/pages/settings/custom-fields/projects/Projects.tsx [hunk 11]
```diff
@@ -16,12 +17,15 @@ import { Field } from '../components';
...
+
   const pages = [
     { name: t('settings'), href: '/settings' },
     { name: t('custom_fields'), href: '/settings/custom_fields' },
     { name: t('projects'), href: '/settings/custom_fields/projects' },
   ];
+
   useEffect(() => {
     document.title = `${import.meta.env.VITE_APP_TITLE}: ${t('custom_fields')}`;
   });
```
Comment (beganovich) [inline src/pages/settings/custom-fields/projects/Projects.tsx:31]: useTitle hook.

File: src/pages/settings/custom-fields/tasks/Tasks.tsx [hunk 12]
```diff
@@ -16,12 +17,15 @@ import { Field } from '../components';
...
+
   const pages = [
     { name: t('settings'), href: '/settings' },
     { name: t('custom_fields'), href: '/settings/custom_fields' },
     { name: t('tasks'), href: '/settings/custom_fields/tasks' },
   ];
+
   useEffect(() => {
     document.title = `${import.meta.env.VITE_APP_TITLE}: ${t('custom_fields')}`;
   });
```
Comment (beganovich) [inline src/pages/settings/custom-fields/tasks/Tasks.tsx:31]: useTitle hook.

File: src/pages/settings/custom-fields/users/Users.tsx [hunk 13]
```diff
@@ -16,12 +17,15 @@ import { Field } from '../components';
...
+
   const pages = [
     { name: t('settings'), href: '/settings' },
     { name: t('custom_fields'), href: '/settings/custom_fields' },
     { name: t('users'), href: '/settings/custom_fields/users' },
   ];
+
   useEffect(() => {
     document.title = `${import.meta.env.VITE_APP_TITLE}: ${t('custom_fields')}`;
   });
```
Comment (beganovich) [inline src/pages/settings/custom-fields/users/Users.tsx:31]: useTitle hook.

File: src/pages/settings/custom-fields/vendors/Vendors.tsx [hunk 14]
```diff
@@ -16,12 +17,15 @@ import { Field } from '../components';
...
+
   const pages = [
     { name: t('settings'), href: '/settings' },
     { name: t('custom_fields'), href: '/settings/custom_fields' },
     { name: t('vendors'), href: '/settings/custom_fields/vendors' },
   ];
+
   useEffect(() => {
     document.title = `${import.meta.env.VITE_APP_TITLE}: ${t('custom_fields')}`;
   });
```
Comment (beganovich) [inline src/pages/settings/custom-fields/vendors/Vendors.tsx:31]: useTitle hook.

File: src/pages/settings/user/components/CustomFields.tsx [hunk 15]
```diff
@@ -22,17 +23,21 @@ export function CustomFields() {
...
+
+      <Card title={t('custom_fields')}>
+        {company &&
+          ['user1', 'user2', 'user3', 'user4'].map((field) => (
+            <Field
+              key={field}
+              initialValue={company.custom_fields[field]}
+              field={field}
+              placeholder={t('user_field')}
+              onChange={(value: any) => handleCustomFieldChange(field, value)}
```
Comment (beganovich) [inline src/pages/settings/user/components/CustomFields.tsx:37]: Is it possible to properly cast this instead of `(value: any)`?

**Review by beganovich — APPROVED** [review]
Looks good, small changes are needed.

File: src/components/CustomFieldsPlanAlert.tsx (hunk 1 above)
Comment (Civolilah) [reply src/components/CustomFieldsPlanAlert.tsx:31]: Sure, just changed!

File: src/pages/invoices/common/components/InvoiceFooter.tsx (hunk 5 above)
Comment (Civolilah) [reply src/pages/invoices/common/components/InvoiceFooter.tsx:60]: Moved

File: src/pages/products/edit/ProductFields.tsx (hunk 3 above)
Comment (Civolilah) [reply src/pages/products/edit/ProductFields.tsx:35]: Moved

File: src/pages/payments/edit/PaymentFields.tsx (hunk 2 above)
Comment (Civolilah) [reply src/pages/payments/edit/PaymentFields.tsx:76]: I discussed with David on the Slack for this one, and he suggested me to completely remove this. So it's removed.

File: src/pages/products/edit/ProductFields.tsx (hunk 4 above)
Comment (Civolilah) [reply src/pages/products/edit/ProductFields.tsx:79]: Same as above!

File: src/pages/settings/custom-fields/components/Field.tsx (hunk 6 above)
Comment (Civolilah) [reply src/pages/settings/custom-fields/components/Field.tsx:43]: Moved

File: src/pages/settings/custom-fields/invoices/Invoices.tsx (hunk 7 above)
Comment (Civolilah) [reply src/pages/settings/custom-fields/invoices/Invoices.tsx:27]: Moved

File: src/pages/settings/custom-fields/invoices/Invoices.tsx (hunk 8 above)
Comment (Civolilah) [reply src/pages/settings/custom-fields/invoices/Invoices.tsx:36]: Sure, refactored!

File: src/pages/settings/custom-fields/payments/Payments.tsx (hunk 9 above)
Comment (Civolilah) [reply src/pages/settings/custom-fields/payments/Payments.tsx:31]: Refactored!

File: src/pages/settings/custom-fields/products/Products.tsx (hunk 10 above)
Comment (Civolilah) [reply src/pages/settings/custom-fields/products/Products.tsx:31]: Refactored

File: src/pages/settings/custom-fields/projects/Projects.tsx (hunk 11 above)
Comment (Civolilah) [reply src/pages/settings/custom-fields/projects/Projects.tsx:31]: refactored

File: src/pages/settings/custom-fields/tasks/Tasks.tsx (hunk 12 above)
Comment (Civolilah) [reply src/pages/settings/custom-fields/tasks/Tasks.tsx:31]: refactored

File: src/pages/settings/custom-fields/users/Users.tsx (hunk 13 above)
Comment (Civolilah) [reply src/pages/settings/custom-fields/users/Users.tsx:31]: Refactored

File: src/pages/settings/custom-fields/vendors/Vendors.tsx (hunk 14 above)
Comment (Civolilah) [reply src/pages/settings/custom-fields/vendors/Vendors.tsx:31]: Refactored

File: src/pages/settings/user/components/CustomFields.tsx (hunk 15 above)
Comment (Civolilah) [reply src/pages/settings/user/components/CustomFields.tsx:37]: Type specification is not needed in this case, the function which accept value is specificed it as string.

**Comment by Civolilah** [comment]
@beganovich Refactoring is done.

**Review by beganovich — APPROVED** [review]
Looks good, thanks :+1:

**Pull request merged**